A71-xds-fallback.md

```diff
@@ -0,0 +1,133 @@
+can be configured using this protocol.
+
+Components of the server mesh communicate with xDS server(s) (also known as
+a "control plane") to obtain the initial configuration and to receive updates
+to the configuration.
+
+An xDS configuration consists of a number of "resources" that can be queried
+separately and may change independently from one another.
+
+Even a brief downtime of the xDS control plane may cause significant disruption
+in the service mesh inter-component connectivity and result in wider outages.
+Examples include:
+
+* Services becoming unavailable as they fail to access downstream dependencies.
+* Traffic being routed incorrectly failing to properly balance the load between
+the endpoints.
+* Security risks may be introduced as outdated credentials are used for
+communications between the servers.
+
+Current xDS implementation in gRPC uses a single global instance of
+the XdsClient class to communicate with the xDS control plane. This instance
+is shared across all channels to reduce overhead by providing a shared cache
+for xDS resources as well as reducing a number of connections to xDS servers.
+
+### Related Proposals: 
+* [A27: xDS-Based Global Load Balancing](A27)
+* [A36: xDS-Enabled Servers](A36)
+* [A57: XdsClient Failure Mode Behavior](A57)
+
+[A27]: A27-xds-global-load-balancing.md
+[A36]: A36-xds-for-servers.md
+[A57]: A57-xds-client-failure-mode-behavior.md
+
+## Proposal
+
+gRPC code should use fallback iff both:
+
+* primary server not reachable
+* at least one watcher exists for a resource that is not cached (where
+   "does not exist" counts as cached).
+
+Instead of using a global XdsClient instance, gRPC will use a shared XdsClient
+instance for each data plane target.  In other words, if two channels are
+created for the target "xds:foo", they will share one XdsClient instance, but
+if another channel is created for "xds:bar", it will use a different XdsClient
+instance.
+
+The following changes will be made to the codebase:
+
+1. Add support for multiple xDS servers in the `bootstrap.json`.
+2. Update implemention to support configuration with multiple xDS servers.
+3. Refactor XdsClient to support multiple instances.
+4. Update code relying on XdsClient to work with multiple instances
+    * xDS resource fetching and tracking.
+    * xDS stats *[???] (do we want to collect them cross-XdsClients?)*
+
+#### bootstrap.json
+
+Currently `bootstrap.json` supports multiple xDS servers but semantics are
+not explicitely specified.
+
+1. xDS servers will be attempted in the order they are specified in
+the bootstrap.json. Server will only be attempted if the previous entry in
+the list is not available.
+1. xDS client will report a failure if the last entry in the list is not
+available.
+1. `channel-creds` or any other server attributes are not shared and need
+to be defined independently for every server.
+
+*[???] Are there any cases when we revert to a primary server?*
```

Comment:
@markdroth my understanding it is a one-way trip. Am I right?

Reply:
Why would switching from server 2 to server 3 be different than switching from server 2 back to server 1 (assuming the ordering is 1, 2, 3)?

Reply:
@larry-safran primary server is preferred and has higher quality data.

@markdroth what happens when we have primary/fallback1/fallback2 servers and we end up on fallback2 and then `fallback1` comes back online?

Reply:
I think this mechanism should work for fallback from server number N to server number N+1, for any non-negative integer value of N.

If server 0 fails, we go to server 1.  If server 1 fails, we go to server 2.

At any point, if server N recovers after having been unreachable, then we stop using server M where M > N.